### PR TITLE
Created working example SFTP server

### PR DESCRIPTION
**Pull request by kothar**
Purely gluing together the standalone_server and the SSH server examples, since
it took me a while to work out how they fit together. Thought it may save the next
person a few minutes!

### CODE REVIEW

File: examples/sftp-server/main.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+	"os"
+
+	"github.com/pkg/sftp"
+	"golang.org/x/crypto/ssh"
+)
+
+// Based on example server code from golang.org/x/crypto/ssh and server_standalone
+func main() {
+
+	readOnly := false
```
Comment (davecheney) [inline examples/sftp-server/main.go]: try
 var (
   readOnly bool
   debufLevelStr string
   ...
)

These will then have their zero value, which will then be overwritten by flag.Parse

File: examples/sftp-server/main.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+	debugStderr := false
+	rootDir := ""
+
+	flag.BoolVar(&readOnly, "R", false, "read-only server")
+	flag.BoolVar(&debugStderr, "e", false, "debug to stderr")
+	flag.StringVar(&debugLevelStr, "l", "none", "debug level")
+	flag.StringVar(&debugLevelStr, "root", "", "root directory")
+	flag.Parse()
+
+	// term := terminal.NewTerminal(channel, "> ")
```
Comment (davecheney) [inline examples/sftp-server/main.go]: delete

File: examples/sftp-server/main.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+			if c.User() == "testuser" && string(pass) == "tiger" {
+				return nil, nil
+			}
+			return nil, fmt.Errorf("password rejected for %q", c.User())
+		},
+	}
+
+	privateBytes, err := ioutil.ReadFile("id_rsa")
+	if err != nil {
+		panic("Failed to load private key")
```
Comment (davecheney) [inline examples/sftp-server/main.go]: log.Fatal is better than panic in examples

File: examples/sftp-server/main.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	private, err := ssh.ParsePrivateKey(privateBytes)
+	if err != nil {
+		panic("Failed to parse private key")
```
Comment (davecheney) [inline examples/sftp-server/main.go]: same

File: examples/sftp-server/main.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	config.AddHostKey(private)
+
+	// Once a ServerConfig has been configured, connections can be
+	// accepted.
+	listener, err := net.Listen("tcp", "0.0.0.0:2022")
+	if err != nil {
+		panic("failed to listen for connection")
```
Comment (davecheney) [inline examples/sftp-server/main.go]: same

File: examples/sftp-server/main.go [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+	}
+	fmt.Printf("Listening on %v\n", listener.Addr())
+
+	nConn, err := listener.Accept()
+	if err != nil {
+		panic("failed to accept incoming connection")
```
Comment (davecheney) [inline examples/sftp-server/main.go]: same,

**Comment by davecheney** [comment]
Looks pretty good to me, thank you. If you can clean up the minor comments I'll merge this for you.

File: examples/sftp-server/main.go (hunk 6 above)
Comment (kothar) [inline examples/sftp-server/main.go]: This code is based on https://godoc.org/golang.org/x/crypto/ssh#example-NewServerConn - is it worth making these suggestions for inclusion there too?

I agree that an example should demonstrate best-practice, especially if it's likely people will use it as the base for future work.

Comment (davecheney) [inline examples/sftp-server/main.go]: Yes please! Old examples age like fruit in the sun. If you could please fix
up those as well that would be great.

**Comment by mdlayher** [comment]
Thanks for this example, it was very helpful!  I look forward to seeing it merged.

**Comment by davecheney** [comment]
@mdlayher if you want to take over this PR that would be great.

**Comment by mdlayher** [comment]
@davecheney Sure, would you just like the other examples updated as well?  I see your previous comments have been resolved.

**Comment by davecheney** [comment]
Oh, I didn't realise the OP had updated it. I'll merge this now.

**Pull request merged**